File: Problemset/0349-Intersection of Two Arrays/49.GroupAnagram.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

char*** groupAnagrams(char** strs, int strsSize, int* returnSize, int** returnColumnSizes) {
  char ***returnArray;
  int i, j, w;
  int p1=0;
  int posicion1ArrayReturn = 0, posicion2ArrayReturn = 0;
  int coinciden;

  int **hashs=calloc(strsSize, sizeof(int*));
  for (i = 0;i<strsSize; i++) {
    hashs[i]=calloc(26, sizeof(int*));
  }
  for (i=0; i<strsSize; i++) { 
    for (j=0; strs[i][j]!='\0'; j++) {
      hashs[i][strs[i][j]-'a']++;
    }
  }
  //hashs[4]=calloc(26, sizeof(int*));
  //ver tabla de hashs
  
  for (i=0; i<strsSize; i++) { 
    printf("%d\t", i);
    for (j=0; j<26; j++) {
      printf("%d",hashs[i][j]);
    }
    printf("\n");
  }

  for (w=0; w<strsSize; w++) {
    for (i=0;i<strsSize;i++){
      coinciden=1;
      printf("pasa %d\n", w);
      if(w==i)
        continue;
      for (j=0; j<26; j++) {
        if(hashs[i][j]!=hashs[w][j]){
          //printf("Es distinto en la posicion %d %d %d\n", w, i, j);
          coinciden=0;
          break;
        }
      }
      if (coinciden) {
         
      }
    }
    //p1++;
  }
  //printf("%d", *returnColumnSizes[0]);
  return returnArray;
}

int main(){

  char* strs[] = {"eat", "tea", "tan", "ate", "nat", "bat"};
  int strsSize = 6;
  int returnSize;
  int* returnColumnSizes;
  
  groupAnagrams(strs, strsSize, &returnSize, &returnColumnSizes);
  return (1);
}

File: Problemset/0349-Intersection of Two Arrays/49.GroupAnagram.dart
List<List<String>> groupAnagrams(List<String> strs) {
  List<List<String>> listareturn = [];
  List<String> strsordenada = List.from(strs);

  // Ordenar las letras de cada palabra
  for (int i = 0; i < strsordenada.length; i++) {
    List<String> letras = strsordenada[i].split('');
    letras.sort();
    strsordenada[i] = letras.join('');
  }

  List<bool> usada = List.filled(strs.length, false);

  for (int i = 0; i < strs.length; i++) {
    if (usada[i]) continue;

    // Crear nuevo grupo
    listareturn.add([strs[i]]);
    usada[i] = true;

    // Buscar anagramas de strs[i]
    for (int j = i + 1; j < strs.length; j++) {
      if (!usada[j] && strsordenada[i] == strsordenada[j]) {
        listareturn.last.add(strs[j]);
        usada[j] = true;
      }
    }
  }

  return listareturn;
}

void main() {
  List<String> lista = ["eat", "tea", "tan", "ate", "nat", "bat"];
  List<List<String>> retorno = groupAnagrams(lista);
  print(retorno);
}
